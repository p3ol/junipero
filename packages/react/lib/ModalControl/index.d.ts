import { ReactNode, MutableRefObject, ComponentPropsWithoutRef } from 'react';

export declare type ModalControlRef = {
  isJunipero: boolean;
  close(): void;
  open(): void;
  toggle(): void;
  modalRef: MutableRefObject<any>;
};

export declare interface ModalControlProps
  extends ComponentPropsWithoutRef<any> {
  children?: ReactNode | JSX.Element;
  ref?: MutableRefObject<ModalControlRef | undefined>;
}

declare function ModalControl(props: ModalControlProps):
  ReactNode | JSX.Element;

export default ModalControl;
