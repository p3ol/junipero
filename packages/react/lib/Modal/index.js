import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { createPortal } from 'react-dom';
import PropTypes from 'prop-types';
import { classNames, mockState, ensureNode } from '@junipero/core';

import { Remove } from '../icons';

const Modal = forwardRef(({
  animate,
  children,
  className,
  apparition,
  container,
  disabled,
  opened,
  closable = true,
  onToggle,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const contentRef = useRef();
  const wrapperRef = useRef();
  const closeButtonRef = useRef();
  const [state, dispatch] = useReducer(mockState, {
    opened: opened ?? false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    contentRef,
    wrapperRef,
    closeButtonRef,
    opened: state.opened,
    open,
    close,
    toggle,
    isJunipero: true,
  }));

  useEffect(() => {
    dispatch({ opened: disabled ? false : !!opened });
  }, [disabled, opened]);

  const onBackdropClick = e => {
    if (
      disabled ||
      !closable ||
      (e.target !== wrapperRef.current && e.target !== closeButtonRef?.current)
    ) {
      return;
    }

    e.preventDefault();
    close();
  };

  const open = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: true });
    onToggle?.({ opened: true });
  };

  const close = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: false });
    onToggle?.({ opened: false });
  };

  const toggle = () => {
    if (disabled) {
      return;
    }

    state.opened = !state.opened;
    dispatch({ opened: state.opened });
    onToggle?.({ opened: state.opened });
  };

  const Wrapper = ({ transitionState }) => (
    <div
      { ...rest }
      ref={innerRef}
      className={classNames(
        'junipero',
        'modal',
        {
          opened: state.opened,
        },
        className,
      )}
    >
      <div
        tabIndex={0}
        onClick={onBackdropClick}
        ref={wrapperRef}
        className="wrapper"
      >
        <div ref={contentRef} className="content">
          { closable && (
            <a
              href="#"
              ref={closeButtonRef}
              className="close"
              onClick={onBackdropClick}
            >
              <Remove />
            </a>
          ) }
          {(!transitionState || ['entered', 'exiting'].find(
            state => state === transitionState)
          ) &&
            children
          }
        </div>
      </div>
    </div>
  );
  Wrapper.propTypes = {
    transitionState: PropTypes.string,
  };
  const content = animate
    ? animate(Wrapper, { opened: state.opened }) : <Wrapper />;

  return state.opened || animate || apparition === 'css'
    ? container ? createPortal(content, ensureNode(container)) : content
    : null;
});

Modal.displayName = 'Modal';
Modal.propTypes = {
  animate: PropTypes.func,
  apparition: PropTypes.string,
  container: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.object,
    PropTypes.func,
  ]),
  closable: PropTypes.bool,
  disabled: PropTypes.bool,
  onToggle: PropTypes.func,
  opened: PropTypes.bool,
};

export default Modal;
