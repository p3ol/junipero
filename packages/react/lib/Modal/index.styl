@require '../theme/colors'

.junipero.modal
  --background-color: rgba($color-midnight, .9)

  position: fixed
  top: 0
  right: 0
  bottom: 0
  left: 0
  overflow: auto
  display: block

  &.apparition-css
    display: none

    &.opened
      display: block

  & > .wrapper
    width: 100%
    height: 100%
    position: absolute
    z-index: 2
    padding: 50px
    background: var(--background-color)
    overflow: auto
    display: flex
    align-items: center
    justify-content: center

    & > .content
      width: 700px
      max-width: 100%
      background: $color-white
      min-height: 500px
      position: relative
      z-index: 3
      border: none
      border-radius: 4px
      margin: 0 auto
      padding: 20px

      & > a.close
        display: inline-block
        width: 15px
        height: 15px
        position: absolute
        cursor: pointer
        top: 10px
        right: 10px

        &:before, &:after
          content: ''
          height: 100%
          position: absolute
          width: 2px
          background: $color-shuttle-gray
          left: calc(50% - 1px)
          top: 0

        &:before
          transform: rotate(45deg)

        &:after
          transform: rotate(-45deg)

  &.panel
    & > .wrapper
      padding: 0
      justify-content: flex-end
      overflow: hidden

      & > .content
        height: 100%
        overflow: auto
        margin: 0
        border-radius: 0
