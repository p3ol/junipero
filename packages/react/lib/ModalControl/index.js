import { forwardRef, useCallback, useImperativeHandle, useRef } from 'react';

import { ModalContext } from '../contexts';

const ModalControl = forwardRef(({ children }, ref) => {
  const modalRef = useRef();

  useImperativeHandle(ref, () => ({
    modalRef,
    open,
    close,
    toggle,
    isJunipero: true,
  }));

  const open = useCallback(() => {
    modalRef.current?.open();
  }, [modalRef]);

  const close = useCallback(() => {
    modalRef.current?.close();
  }, [modalRef]);

  const toggle = useCallback(() => {
    modalRef.current?.toggle();
  }, [modalRef]);

  const setRef = useCallback(r => {
    modalRef.current = r;
  }, [modalRef]);

  const getContext = useCallback(() => ({
    open,
    close,
    toggle,
    setRef,
  }), [open, close, toggle, setRef]);

  return (
    <ModalContext.Provider value={getContext()}>
      { children }
    </ModalContext.Provider>
  );
});

ModalControl.displayName = 'ModalControl';

export default ModalControl;
