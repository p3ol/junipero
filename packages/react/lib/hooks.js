import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
