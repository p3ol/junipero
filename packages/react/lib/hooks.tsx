import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
