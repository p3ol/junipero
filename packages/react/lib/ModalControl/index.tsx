import { forwardRef, useCallback, useImperativeHandle, useRef, ReactNode, MutableRefObject, ComponentPropsWithoutRef } from 'react';

import { ModalContext } from '../contexts';
import { ModalRef } from '../Modal';

export declare type ModalControlRef = {
  isJunipero: boolean;
  close(): void;
  open(): void;
  toggle(): void;
  modalRef: MutableRefObject<ModalRef>;
};

export declare interface ModalControlProps
  extends ComponentPropsWithoutRef<any> {
  children?: ReactNode | JSX.Element;
  ref?: MutableRefObject<ModalControlRef | undefined>;
}

const ModalControl = forwardRef(({ children }: ModalControlProps, ref) => {
  const modalRef = useRef<ModalRef>();

  useImperativeHandle(ref, () => ({
    modalRef,
    open,
    close,
    toggle,
    isJunipero: true,
  }));

  const open = useCallback(() => {
    modalRef.current?.open();
  }, [modalRef]);

  const close = useCallback(() => {
    modalRef.current?.close();
  }, [modalRef]);

  const toggle = useCallback(() => {
    modalRef.current?.toggle();
  }, [modalRef]);

  const setRef = useCallback((ref: ModalRef) => {
    modalRef.current = ref;
  }, [modalRef]);

  const getContext = useCallback(() => ({
    open,
    close,
    toggle,
    setRef,
  }), [open, close, toggle, setRef]);

  return (
    <ModalContext.Provider value={getContext()}>
      { children }
    </ModalContext.Provider>
  );
});

ModalControl.displayName = 'ModalControl';

export default ModalControl;
