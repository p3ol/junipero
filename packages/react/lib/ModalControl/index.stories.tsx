import { useRef } from 'react';
import { action } from 'storybook/actions';

import { useModal } from '../hooks';
import Modal, { type ModalRef } from '../Modal';
import Button from '../Button';
import ModalControl from './index';

export default { title: 'react/ModalControl' };

const Comp = () => {
  const { open } = useModal();

  return (
    <Button className="primary" onClick={open}>
      Open me
    </Button>
  );
};

export const Basic = () => {
  const modalRef = useRef<ModalRef>(null);

  return (
    <ModalControl>
      <Comp />
      <Modal ref={modalRef} onToggle={action('toggle')}>Test</Modal>
    </ModalControl>
  );
};
