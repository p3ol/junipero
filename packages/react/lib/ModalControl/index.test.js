import { render, fireEvent, screen } from '@testing-library/react';

import { useModal } from '../hooks';
import Modal from '../Modal';
import ModalControl from '.';

describe('<ModalControl />', () => {
  const Comp = () => {
    const { open } = useModal();

    return (
      <a onClick={open}>
        Open me
      </a>
    );
  };

  it('should should allow to control a modal from another component', () => {
    const { container, unmount } = render(
      <div className="container">
        <ModalControl>
          <Comp />
          <Modal container=".container" />
        </ModalControl>
      </div>
    );

    expect(container).toMatchSnapshot('Closed');

    fireEvent.click(screen.getByText('Open me'));

    expect(container).toMatchSnapshot('Opened');

    unmount();
  });
});
