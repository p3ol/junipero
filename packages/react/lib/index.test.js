import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
