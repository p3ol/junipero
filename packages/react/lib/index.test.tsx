import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
