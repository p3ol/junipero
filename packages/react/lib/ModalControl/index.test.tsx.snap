// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<ModalControl /> should should allow to control a modal from another component: Closed 1`] = `
<div>
  <div
    class="container"
  >
    <a>
      Open me
    </a>
  </div>
</div>
`;

exports[`<ModalControl /> should should allow to control a modal from another component: Opened 1`] = `
<div>
  <div
    class="container"
  >
    <a>
      Open me
    </a>
    <div
      class="junipero modal opened"
    >
      <div
        class="wrapper"
        tabindex="0"
      >
        <div
          class="content"
        >
          <a
            class="close"
            href="#"
          >
            <svg
              class="junipero icon remove"
              height="10"
              viewBox="0 0 9 10"
              width="10"
            >
              <path
                d="M8 1.5L1 8.5"
              />
              <path
                d="M1 1.5L8 8.5"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
